# Data Cleaning

## Import Relevant Packages and Libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandoc

In [3]:
pd.set_option('display.max_columns', None)

## Import the data

In [4]:
# read in the data
df_insta = pd.read_csv('../data/raw/Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_Instagram.csv')
df_lnkd = pd.read_csv('../data/raw/Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_Linkedin.csv')
df_twit = pd.read_csv('../data/raw/Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_Twitter.csv')
df_fbk = pd.read_csv('../data/raw/Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_Facebook.csv')

C:\Users\Victor Essien\AppData\Local\Temp\ipykernel_2536\2350594635.py:2: DtypeWarning: Columns (21,22,30,31,32,33,40,78,80) have mixed types. Specify dtype option on import or set low_memory=False.
  df_insta = pd.read_csv('../data/raw/Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_Instagram.csv')
C:\Users\Victor Essien\AppData\Local\Temp\ipykernel_2536\2350594635.py:3: DtypeWarning: Columns (78,80) have mixed types. Specify dtype option on import or set low_memory=False.
  df_lnkd = pd.read_csv('../data/raw/Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_Linkedin.csv')
C:\Users\Victor Essien\AppData\Local\Temp\ipykernel_2536\2350594635.py:4: DtypeWarning: Columns (32,33,41,53,61,78,146) have mixed types. Specify dtype option on import or set low_memory=False.
  df_twit = pd.read_csv('../data/raw/Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_Twitter.csv')
C:\Users\Victor Essien\AppData\Local\Temp\ipykernel_2536\2350594635.py:5: DtypeW

## First Look

In [5]:
# function to get the shape of all dataframes
def get_shape(*dataframes):
    for idx, df in enumerate(dataframes, start=1):
        print(f"Shape of {df['Network'].values[0]} dataset: {df.shape}")
    

In [6]:
get_shape(df_insta, df_lnkd, df_twit, df_fbk)

Shape of Instagram dataset: (10000, 147)
Shape of LinkedIn dataset: (7760, 147)
Shape of Twitter dataset: (8529, 147)
Shape of Facebook dataset: (9803, 147)


They all have the same number of columns, but are they the same?

In [7]:
# check if all dataframes have the same columns
if list(df_insta.columns) == list(df_twit.columns) == list(df_lnkd.columns) == list(df_fbk.columns):
   print('True')
else: 
    print('False')

True


Since they all have the same columns, a function will be needed to clean all of them. Before that the columns need to be investigated to determine what impurities or inconsistencies exist within the data, so as to inform cleaning steps.

## Cleaning

To clean all dataframes without concatenating them, I create a test dataframe, scope the columns for impurities and inconsistencies then define a function to generalize to the four dataframes.

### Defining the cleaning process:
- Standardize column names
- Set appropriate data types
- Check for general inconsistencies
- Check for and deal with missing values
- Remove duplicates
- Remove excessive whitespace from string columns
- Define function and Save clean data to a csv file

#### Standardize column names

In [8]:
# make a copy of one dataset
df_test = df_insta.copy()

In [9]:
# view columns
df_test.columns

Index(['Date', 'Post ID', 'Network', 'Post Type', 'Content Type', 'Profile',
       'Sent by', 'Link', 'Post', 'Linked Content',
       ...
       'Video Removed from Playlists', 'Annotation Impressions',
       'Annotation Clickable Impressions', 'Annotation Closable Impressions',
       'Annotation Closes', 'Card Impressions', 'Card Teaser Impressions',
       'Card Teaser Clicks', 'Poll Votes', 'Tags'],
      dtype='object', length=147)

In [10]:
# replace spaces and special character with '_'
df_test.columns = df_test.columns.str.lower().str.replace('-', '_').str.replace(' ', '_').str.replace('(', '').str.replace(')', '')

#### Set appropriate data types

To do this, the columns need to be checked to understand why pandas could not define their data types on import or what data types are mixed within such columns.

In [11]:
df_test.dtypes

date                        object
post_id                      int64
network                     object
post_type                   object
content_type                object
                            ...   
card_impressions           float64
card_teaser_impressions    float64
card_teaser_clicks         float64
poll_votes                 float64
tags                        object
Length: 147, dtype: object

From the output above of the data types of each column, it can be seen already that some columns dont have the correct data type. This may be due to inconsistencies of the data within those columns. So those columns need to be cleaned and set to the appropriate data type. An example is the date column.

A further look can be had at the data type of the first 20 columns below:

In [12]:
df_test.dtypes[:20]

date                            object
post_id                          int64
network                         object
post_type                       object
content_type                    object
profile                         object
sent_by                         object
link                            object
post                            object
linked_content                 float64
impressions                     object
organic_impressions             object
viral_impressions              float64
non_viral_impressions          float64
paid_impressions               float64
fan_impressions                float64
fan_organic_impressions        float64
fan_paid_impressions           float64
non_fan_impressions            float64
non_fan_organic_impressions    float64
dtype: object

The impressions columns should contain only numbers.

In [13]:
 # convert date column to datetime
df_test['date'] = pd.to_datetime(df_test['date'], format='%m/%d/%Y %I:%M %p')
       
# convert string columns with numbers to float
for column in df_test.columns:
    if df_test[column].dtype == 'object':
        if df_test[column].str.isnumeric().any():
            df_test[column] = df_test[column].str.replace(',','').astype(float)
        elif 'per' in df_test[column].name:
            df_test[column] = df_test[column].str.replace(',','').str.replace('%','').astype(float)
        

In [14]:
# check to confirm 
df_test.dtypes[:20]

date                           datetime64[ns]
post_id                                 int64
network                                object
post_type                              object
content_type                           object
profile                                object
sent_by                                object
link                                   object
post                                   object
linked_content                        float64
impressions                           float64
organic_impressions                   float64
viral_impressions                     float64
non_viral_impressions                 float64
paid_impressions                      float64
fan_impressions                       float64
fan_organic_impressions               float64
fan_paid_impressions                  float64
non_fan_impressions                   float64
non_fan_organic_impressions           float64
dtype: object

#### Check for general inconsistencies

While scoping some of the categorical columns, something curious was discovered. See below:

In [15]:
df_test.sent_by.value_counts()

sent_by
                     4967
Blessing Ubah        2707
Aramide Salami       1733
Kemi Amoo             315
Sophia Amuka          192
Damilare Oyekanmi      86
Name: count, dtype: int64

There are 4967 post made by ' '. There are too many rows to just drop and those rows contain important data from which insights can be extracted. The best thing is to replace that white space with 'admin'. The reasoning is someone was making those posts on behalf of the profile.

In [16]:
# replace white space in sent_by with admin
df_test['sent_by'] = df_test['sent_by'].replace(' ', 'admin')

In [17]:
# to confirm
df_test.sent_by.value_counts()

sent_by
admin                4967
Blessing Ubah        2707
Aramide Salami       1733
Kemi Amoo             315
Sophia Amuka          192
Damilare Oyekanmi      86
Name: count, dtype: int64

#### Check for and deal with missing values

In [18]:
# check percentage of missing values per column
missing_percentage = df_test.isnull().sum()/len(df_test)*100
missing_types = df_test.dtypes
missing_values = pd.DataFrame({'missing_percentage' : missing_percentage, 'dtype' : missing_types})

In [19]:
# missing_values.missing_percentage *10000
missing_values[:20]

missing_percentage           dtype
date                                       0.00  datetime64[ns]
post_id                                    0.00           int64
network                                    0.00          object
post_type                                  0.00          object
content_type                               0.00          object
profile                                    0.00          object
sent_by                                    0.00          object
link                                       0.00          object
post                                       0.11          object
linked_content                           100.00         float64
impressions                               14.84         float64
organic_impressions                       14.84         float64
viral_impressions                        100.00         float64
non_viral_impressions                    100.00         float64
paid_impressions                         100.00         float64
fan_impressions                          100.00         float64
fan_organic_impressions                  100.00         float64
fan_paid_impressions                     100.00         float64
non_fan_impressions                      100.00         float64
non_fan_organic_impressions              100.00         float64

In [20]:
df_test[df_test.post.isnull()]

date            post_id    network post_type content_type  \
1423 2021-10-06 15:18:00  18264735889006746  Instagram      Post        Video   
2687 2022-11-07 15:15:00  17996455921508611  Instagram      Post        Video   
3037 2021-06-02 14:36:00  17928942340583441  Instagram      Post        Video   
3466 2019-11-10 10:22:00  18075127972178513  Instagram      Post        Photo   
4057 2021-08-05 14:36:00  17877664865423804  Instagram      Post        Video   
4680 2021-12-08 14:55:00  17947547986621047  Instagram      Post        Video   
4721 2021-11-10 14:43:00  17916752206969560  Instagram      Post        Video   
4881 2021-12-17 15:39:00  17955736783523769  Instagram      Post        Video   
5297 2021-12-17 15:33:00  18119747119255695  Instagram      Post        Video   
8542 2022-07-08 10:05:00  17937413156227992  Instagram      Post     Carousel   
9343 2015-12-02 20:50:00  17843461603092131  Instagram      Post        Video   

          profile       sent_by                                       link  \
1423  stanbicibtc         admin  https://www.instagram.com/tv/CUsLOdsoeK7/   
2687  stanbicibtc         admin  https://www.instagram.com/tv/Ckqad7iI4Lx/   
3037  stanbicibtc         admin  https://www.instagram.com/tv/CPnqTksg0xY/   
3466  stanbicibtc         admin   https://www.instagram.com/p/B4rgFtLFS0Y/   
4057  stanbicibtc         admin  https://www.instagram.com/tv/CSMdKslodw-/   
4680  stanbicibtc         admin  https://www.instagram.com/tv/CXOWnewoDbO/   
4721  stanbicibtc         admin  https://www.instagram.com/tv/CWGPHnaIN6D/   
4881  stanbicibtc         admin  https://www.instagram.com/tv/CXlm5xEoRdP/   
5297  stanbicibtc         admin  https://www.instagram.com/tv/CXlmQIroVuz/   
8542  stanbicibtc  Sophia Amuka   https://www.instagram.com/p/CfvuCJTsCuV/   
9343  stanbicibtc         admin    https://www.instagram.com/p/-zUIMjyvZ1/   

     post  linked_content  impressions  organic_impressions  \
1423  NaN             NaN       3024.0               3024.0   
2687  NaN             NaN       2195.0               2195.0   
3037  NaN             NaN       2051.0               2051.0   
3466  NaN             NaN       1823.0               1823.0   
4057  NaN             NaN       1469.0               1469.0   
4680  NaN             NaN       1193.0               1193.0   
4721  NaN             NaN       1249.0               1249.0   
4881  NaN             NaN       1104.0               1104.0   
5297  NaN             NaN        883.0                883.0   
8542  NaN             NaN          NaN                  NaN   
9343  NaN             NaN          NaN                  NaN   

      viral_impressions  non_viral_impressions  paid_impressions  \
1423                NaN                    NaN               NaN   
2687                NaN                    NaN               NaN   
3037                NaN                    NaN               NaN   
3466                NaN                    NaN               NaN   
4057                NaN                    NaN               NaN   
4680                NaN                    NaN               NaN   
4721                NaN                    NaN               NaN   
4881                NaN                    NaN               NaN   
5297                NaN                    NaN               NaN   
8542                NaN                    NaN               NaN   
9343                NaN                    NaN               NaN   

      fan_impressions  fan_organic_impressions  fan_paid_impressions  \
1423              NaN                      NaN                   NaN   
2687              NaN                      NaN                   NaN   
3037              NaN                      NaN                   NaN   
3466              NaN                      NaN                   NaN   
4057              NaN                      NaN                   NaN   
4680              NaN                      NaN                   NaN   
4721              NaN           

In [21]:
df_test[df_test.impressions.isnull()]

date            post_id    network post_type content_type  \
8516 2023-06-12 13:08:00  18029321227525834  Instagram      Post     Carousel   
8517 2023-06-08 09:55:00  18365455210048288  Instagram      Post        Photo   
8518 2023-06-07 13:30:00  17964499847327992  Instagram      Post        Photo   
8519 2023-05-16 20:30:00  18005669605736790  Instagram      Post        Photo   
8520 2023-04-27 20:18:00  17979979232158293  Instagram      Post        Photo   
...                  ...                ...        ...       ...          ...   
9995 2014-12-24 17:37:00  17841808444024797  Instagram      Post        Photo   
9996 2014-12-24 10:28:00  17841674617024797  Instagram      Post        Photo   
9997 2014-12-23 17:14:00  17842896850024797  Instagram      Post        Photo   
9998 2014-12-23 10:38:00  17841674620024797  Instagram      Post        Photo   
9999 2014-12-22 17:59:00  17841808438024797  Instagram      Post        Photo   

          profile        sent_by                                      link  \
8516  stanbicibtc  Blessing Ubah  https://www.instagram.com/p/CtY8c1Ts0pP/   
8517  stanbicibtc  Blessing Ubah  https://www.instagram.com/p/CtOTPwcMPZ-/   
8518  stanbicibtc      Kemi Amoo  https://www.instagram.com/p/CtMG-nLOyVP/   
8519  stanbicibtc      Kemi Amoo  https://www.instagram.com/p/CsUNo24OrlI/   
8520  stanbicibtc   Sophia Amuka  https://www.instagram.com/p/CrjRFyLByBP/   
...           ...            ...                                       ...   
9995  stanbicibtc          admin   https://www.instagram.com/p/w_xUG5yvRB/   
9996  stanbicibtc          admin   https://www.instagram.com/p/w_AUutyvQD/   
9997  stanbicibtc          admin   https://www.instagram.com/p/w9J5vpyvY4/   
9998  stanbicibtc          admin   https://www.instagram.com/p/w8cnwpSvY5/   
9999  stanbicibtc          admin   https://www.instagram.com/p/w6qRb-yvai/   

                                                   post  linked_content  \
8516  At Stanbic IBTC, we hold special gatherings fo...             NaN   
8517  Motor insurance is required for any ‘Idan’ who...             NaN   
8518  Did you know that you can use your retirement ...             NaN   
8519  We celebrate the boy child today and every day...             NaN   
8520  Make your money bloom with the right investmen...             NaN   
...                                                 ...             ...   
9995  Withdraw cash up to 5 times, pay bills or tran...             NaN   
9996  #CardDiaries week 13\n\nI love Christmas! In m...             NaN   
9997  Eat-in this holidays. \nInstead of going out t...             NaN   
9998  If you do not move forward, you will always be...             NaN   
9999  Travel to Dubai this season with #StanbicDelig...             NaN   

      impressions  organic_impressions  viral_impressions  \
8516          NaN                  NaN                NaN   
8517          NaN                  NaN                NaN   
8518          NaN                  NaN                NaN   
8519          NaN                  NaN                NaN   
8520          NaN                  NaN                NaN   
...           ...                  ...                ...   
9995          NaN                  NaN                NaN   
9996          NaN                  NaN                NaN   
9997          NaN                  NaN                NaN   
9998          NaN                  NaN                NaN   
9999          NaN                  NaN                NaN   

      non_viral_impressions  paid_impressions  fan_impressions  \
8516                    NaN               NaN              NaN   
8517                    NaN               NaN              NaN   
8518                    NaN               NaN              NaN   
8519                    NaN               NaN              NaN   
8520                    NaN               NaN              NaN   
...                     ...               ...              ...   
9995          

Observing the frames above, for both categorical and numerical columns dropping those missing values doesn't make much sense. For the numerical columns however, filling with zero should be done. This will be done during EDA.

#### Remove duplicates

In [22]:
# check for and remove duplicates
if df_test.duplicated().sum() > 0:
    df_test.drop_duplicates(inplace=True)

#### Remove excessive whitespace from string columns

In [23]:
for column in df_test.columns:
    if df_test[column].dtype == 'object':
        df_test[column] = df_test[column].str.strip()

#### Define function and Save clean data to a csv file

In [24]:
def clean_data(*dataframes):
    """
    Cleans and standardizes a list of DataFrames and saves them to CSV files.

    Parameters:
    *dataframes (pd.DataFrame): Variable number of input DataFrames to be cleaned.

    Notes:
    - Column names are converted to lowercase and special characters are replaced with underscores.
    - The 'date' column is converted to a datetime format.
    - Columns containing numeric strings are converted to float, handling comma and percent formatting.
    - 'sent_by' column is corrected by replacing white spaces with 'admin'.
    - Duplicate rows are dropped if any exist.
    - Each cleaned DataFrame is saved to a separate CSV file with a unique name.
    """
    
    for i, dataframe in enumerate(dataframes, 1):
        # standardize column names
        dataframe.columns = dataframe.columns.str.lower()\
        .str.replace('-', '_')\
        .str.replace(' ', '_')\
        .str.replace('(', '')\
        .str.replace(')', '')
        
        # convert date column to datetime
        dataframe['date'] = pd.to_datetime(dataframe['date'], format='%m/%d/%Y %I:%M %p')
       
        # convert string columns with numbers to float
        for column in dataframe.columns:
            if dataframe[column].dtype == 'object':
                if dataframe[column].str.isnumeric().any():
                    dataframe[column] = dataframe[column].str.replace(',','').astype(float)
                elif 'per' in dataframe[column].name:
                    dataframe[column] = dataframe[column].str.replace(',','').str.replace('%','').astype(float)
                elif 'rate' in dataframe[column].name:
                    dataframe[column] = dataframe[column].str.replace(',','').str.replace('%','').astype(float)
        
        # replace white space in sent_by with admin
        dataframe['sent_by'] = dataframe['sent_by'].replace(' ', 'admin')
        
        # drop duplicates if any exist
        if dataframe.duplicated().sum() > 0:
            dataframe.drop_duplicates(inplace=True)
        
        # remove trailing whitespace from string columns
        for column in dataframe.columns:
            if dataframe[column].dtype == 'object':
                dataframe[column] = dataframe[column].str.strip()
        
        # Save the cleaned DataFrame to a CSV file with a unique name
        csv_filename = f"../data/processed/Cleaned-Post Performance (Stanbic IBTC) January 1, 2013 - July 13, 2023_{dataframe['network'].values[0]}.csv"
        dataframe.to_csv(csv_filename, index=False)
        
            
    

In [25]:
clean_data(df_insta, df_lnkd, df_twit, df_fbk)